Altimetrik Week 7 - Salesforce Apex Project

diegoaom opened the pull request:
This was the project for the week 7 of the Altimetrik Become a Salesforce Developer Bootcamp.

Universal Containers Lawyers is a fictional company that provides legal advisory to their customers.
They consume data from an external system that stores the information into Heroku on a daily
basis.
They want to adopt Salesforce and needed to migrate the information stored in Heroku into an org.
For this purpose a Salesforce integration with Heroku Api was made.

I tried my best to achieve the best practices possible or in my knowledge.
10 Apex classes were made, for the purpose of making it in the most modular way possible, respecting the single responsability principle in each method. It wasnt always achieved but i think i did my best to accomplish it.

Where i think there was a big change of improvement is in the testing part, the test done in the app only achieved a mere 83% code coverage, in which most are from failure in testing DML exceptions.

![ucl1](https://github.com/diegoaom/Universal-Containers-Lawyers/assets/54308070/3bef25e5-58ef-4e60-8bcc-fcb5ca2883ea)
![image](https://github.com/diegoaom/Universal-Containers-Lawyers/assets/54308070/b3c58395-17f1-4a81-9883-f5ec828ac218)
![image](https://github.com/diegoaom/Universal-Containers-Lawyers/assets/54308070/a9670367-c4b0-4873-bfe4-a207bfdbdf5d)
![image](https://github.com/diegoaom/Universal-Containers-Lawyers/assets/54308070/28a2c9f9-e089-429a-a369-c843082c5147)
![image](https://github.com/diegoaom/Universal-Containers-Lawyers/assets/54308070/75447ae0-0999-4245-aedd-131b8af03472)

diegoaom committed:
feat: custom object 'Legal Advisor' added.
Legal_advisor__c custom object was created with four fields and one validation rule.
Account Name, Account Number, Account Status as Text Fields.
As of Date as a Date field.
Validation Rule 'EnabledOrDisabledOnly' which only allows 'Enabled' or 'Disabled' Status
in Account Status was added.

diegoaom committed:
feat: custom object 'Client' added.
Client__c custom object was created with six fields and one master-detail relationship with 'Legal Advisor'.
First Name, Last Name, Email, Phone, Address, Client Number as Text Fields.
Master-Detail relationship as detail with 'Legal Advisor' Custom Object added.

diegoaom committed:
feat: wrapper class 'Ucl_LegalAdvisor' added.
'Ucl_legalAdvisor' is a wrapper class used for the REST Api callout, because the structure of the JSON is already known
this class helps with the deserialization of the JSON.
It includes the Legal Advisor entity, the Client entity and a parse method.

diegoaom committed:
feat: 'Ucl_ErrorHandlingUtils' utils class added.
'Ucl_ErrorHandlingUtils' is a utils class which has a method for using the SingleEmailMessage methods.
It's purpose is to send an email to the system administrator when an exception or error happens during the REST Api callout.

diegoaom committed:
feat: 'Ucl_RestCallout' class added.
'Ucl_RestCallout' class has a method that makes a httpRequest to the heroku REST Api and in case of success returns a list of objects already parsed
by the method in 'Ucl_LegalAdvisor'. In case of an exception or error it logs the error and sends an email to the system administrator using
the method in 'Ucl_ErrorHandlingUtils', also returns an empty list to avoid a 'Null' return

diegoaom committed:
refactor: reworked 'Ucl_LegalAdvisor' class.
'Ucl_LegalAdvisor' was remade for a better use of the 'Ucl_RestCallout'.
A constructor method was added and 'Ucl_Client' was added to their own class.
Parse method was removed because it didnt accomplish what was expected.
Summary: 'Ucl_LegalAdvisor' was reworked to better fit the object-oriented paradigm.

diegoaom committed:
feat: 'Ucl_Client' class added.
'Ucl_Client' class has a constructor method, previously it was together with 'Ucl_LegalAdvisor'

diegoaom committed:
refactor: 'Ucl_RestCallout' reworked.
'Ucl_RestCallout' was reworked because it didnt accomplish what was expected.
Previously it returned a list, now it returns a map.
'Ucl_LegalAdvisor' constructor is now used in the getData method.
This change was made to avoid using nested for loops in the insert class.

diegoaom committed:
refactor: Ucl_ErrorHandlingUtils rename to Ucl_EmailNotificationManager.
This change was done for better readability because Error Logs constructor have its own method.

diegoaom committed:
feat: Ucl_ClientDelete added.
This class have a method for deleting clients that are not coming anymore in the callout.
It uses Client's client number and Legal Advisor Account Number concatenated to search for clients that need to delete, thanks to a formula field added in clients object.
It also includes a way to avoid deleting clients made manually by the system admin.

diegoaom committed:
test: Ucl_ClientDelete test added.
dml exception testing wasnt possible to do, so this test doesnt have 100% coverage.

diegoaom committed:
fix: removed explicit bind in Ucl_Client.
removed 'this' from the Ucl_Client constructor.

diegoaom committed:
test: Ucl_Client test added.
100% coverage

diegoaom committed:
feat: Ucl_ErrorLogManager added.
Ucl_ErrorLogManager contains a method for constructing a Error_Log__c and then insert it.
Was created for reusability in between each method.

diegoaom committed:
test: Ucl_ErrorLogManager.
100% coverage

diegoaom committed:
feat: Ucl_ErrorLogsCleanup added.
This is a schedulable class for the purpose of deleting logs that are one month old.
It runs as a apex job daily every midnight.

diegoaom committed:
test: Ucl_ErrorLogsCleanup.
100% coverage

diegoaom committed:
fix: removed explicit binding and formatting.
Removed 'this' bind and formatted join variable for better readability

diegoaom committed:
test: Ucl_legalAdvisor.
100% coverage

diegoaom committed:
fix: added Ucl_ErrorLogManager method and removed debug statements.
Ucl_ErrorLogManager was added.
Debug statements were removed.

diegoaom committed:
test: Ucl_RestCallout
100% coverage

diegoaom committed:
feat: Ucl_ScheduledDataProcessor added.
This is a schedulable class that runs daily at midnight for the purpose of orchestrating
the callout, the upsert, the update and the delete of legal advisors and clients

diegoaom committed:
test: Ucl_ScheduledDataProcessor.
100% coverage

diegoaom committed:
feat: Ucl_UpdateLegalAdvisorStatus added.
This class has a method that takes a map from the callout response and takes every legal advisor with status:
'disabled' and updates the records that have the same legal advisor but with 'enabled' status.

diegoaom committed:
test: Ucl_UpdateLegalAdvisorStatus.
There is no testing for dml exceptions cases

diegoaom committed:
feat: Ucl_UpsertManager added.
This class has two methods: one has the task of upserting legal advisors record.
The other method upsert the clients record.

diegoaom committed:
test: Ucl_UpsertManager.
There is no coverage of DML exceptions.

diegoaom committed:
feat: restCalloutButton added.
This Lightning Web Component is used to invoke the callout method, it shows a toast message with a variant of success if the invokation
is successful and a toast message with the variant of error in case of failure.

diegoaom committed:
feat: added a formula field and a checkbox field in Client
The formula field is used for concatenating legal advisor account number and client client number.
Its purpose is for identify clients.
The checkbox field is for identify clients created by system admins.
Deleted Validation rules for enabled or disabled status as a bugfix.

diegoaom committed:
feat: Error_Log__c custom object added.
It has a Error_Message__c field of string type
Location__c field of string type to know where the error was.
Timestamp__c as datetime type.

diegoaom committed:
fix: added search for Legal_Advisor__c.

diegoaom committed:
utils: json files for mocking in testing.

diegoaom committed:
feat: Client__c custom tab added.

diegoaom committed:
feat: Error_Log__c custom tab added.

diegoaom committed:
feat: restCalloutButton lightning tab added.

diegoaom committed:
feat: Legal_Advisor__c custom tab added.

diegoaom committed:
feat: Universal_Containers_Lawyers_app added.
It contains four tabs: Legal Advisors, Clients, Error Logs and restCalloutButton tab.

diegoaom committed:
fix: wrong json file was commited.

diegoaom committed:
feat: executeScheduledDataProcessor method added.
This method executes the nightly service on demand.

diegoaom committed:
refactor: executeScheduledDataProcessor instead of getData.
getData method was not doing what the task was.
executeScheduledDataProcessor method does the service invocation as expected.

diegoaom committed:
fix: error log now display all fields

jbessonart commented on force-app/main/default/classes/Ucl_EmailNotificationManager.cls:
```diff
@@ -0,0 +1,11 @@
+public with sharing class Ucl_EmailNotificationManager {
+  public static void sendErrorEmail(String subject, String body) {
+    Messaging.SingleEmailMessage email = new Messaging.SingleEmailMessage();
+    String[] toAddresses = new List<String>{ 'diegoolivierivii@gmail.com' };
```
Esto esta bien, asi que no te preocupes, pero a modo de comentario, podrias haber hecho una query a el objeto User where profile = "System Administrator" y agregar sus emails a la lista de "emailAddresses" entonces se lo manda a todos los admin (entre ellos a vos, igual supongo habia 1 admin nomas).

jbessonart commented on force-app/main/default/classes/Ucl_UpsertManager.cls:
```diff
@@ -0,0 +1,81 @@
+  }
+
+    try {
+      upsert clientsToUpsert Client_Number__c;
```
Aca vas a tener un problema, porque los Client_Number__c NO se puede usar como key porque se pueden repetir entre legal advisor. Lo recomendado es crear un nuevo campo que sea "Client_Key__c" o algo asi, que puedas cuando creas el cliente setearlo como : client.Client_Key__c = legalA.Account_Number__c + client.Client_Number__c
entonces si, despues podes hacer un "upsert clientsToUpsert Client_Key__c;" . En tu caso ya te creaste un formula field AccountNumberPlusClientNumber__c, no se si lo podes poner como external id, sino usa un campo nuevo.

diegoaom committed:
rework: AccountNumberPlusClientNumber__c changed to Client_Key__c
AccountNumberPlusClientNumber__c was previously the formula field responsible from concatenating Legal_Advisor__c.Account_Number__c and
Client__c.Client_Number__c, but because it was a formula field, it cant be used as an external id or indexed so it was replaced.
Client_Key__c is the custom field that replaces the formula field and is a text type field.

diegoaom committed:
fix: updated field used.
Because of the change from AccountNumberPlusClientNumber__c to Client_Key__c, both classes need to be updated.

diegoaom committed:
fix: changed client_number__c to Client_Key__c.
Client_Number__c was wrongfully used as an external id in the upsert statement.
This was changed to the new Client_Key__c that uses Legal_Advisor__c.Account_Number__c + Client__c.Client_Number__c.

diegoaom committed:
rework: Ucl_EmailNotificationManager behaviour changes.
Changed Ucl_EmailNotificationManager hardcoded method to a more dynamic one, this allow for a better scaling in case of adding new system administrators.

diegoaom committed:
feat: schedulableDataProcessor method added.
This method was added because previously the schedulable was called in a sincronic way which apex jobs dont allow, which caused DML exceptions
and was making the nightly service not work properly.
This wrapper method is used to make the dataProcessingService method async.

diegoaom committed:
fix: Client_Key__c substring operation added.
Client_Key__c now incorporates a substring to prevent DML exceptions caused by exceeding the allowed length.

diegoaom committed:
feat: Number_of_Clients custom field added.
Number_of_Clients__c field from Legal_Advisor__c object is added.
This field is a rollup field designed to offer insights about how many clients a legal advisor has.
